### DISCUSSION TITLE
document is not defined in backend side (nodejs)

### DISCUSSION THREAD

**Discussion by patrickalima98** [ueberdosis/tiptap#3233] [General Questions] [answered]
Hello how are you!

First of all, thank you so much for this amazing project!

I have a question, I'm developing a backend in which it should be possible to send parts of documents to update a document already created in the database.

I thought about using the Api `setContent` of the Editor instance, but when using the Editor on the backend (nodejs), I get the following error:

```
err: {
       "type": "ReferenceError",
       "message": "document is not defined",
       "stack":
           ReferenceError: document is not defined
```

How can I use an instance of the editor, so that I can insert or remove parts of the document, in the backend with nodejs. I couldn't find any information about it.

thanks in advance

**Comment by paliwalsamarth** [comment]
Little late at the party, I was also trying something similar, I was able to get it working using some browser mock api packages, like below
<code>
const jsdom = require("jsdom");
const { JSDOM } = jsdom;
const dom = new JSDOM(`<html><div class="element"></div></html>`);
export const window = dom.window;
export const document = window.document;
global.document = document;
global.window = window;
</code>
<br>

**Important: In extensions list, if you have Collaboration Extension, then it will give below error, but without it, it works.**

<code>
Error: Unexpected content type in insert operation
at /home/node/app/node_modules/yjs/dist/yjs.cjs:5382:23
at Array.forEach (<anonymous>)
at typeListInsertGenericsAfter (/home/node/app/node_modules/yjs/dist/yjs.cjs:5353:11)
at typeListInsertGenerics (/home/node/app/node_modules/yjs/dist/yjs.cjs:5410:12)
at /home/node/app/node_modules/yjs/dist/yjs.cjs:7737:9
at transact (/home/node/app/node_modules/yjs/dist/yjs.cjs:3415:14)
at YXmlFragment.insert (/home/node/app/node_modules/yjs/dist/yjs.cjs:7736:7)
at file:///home/node/app/node_modules/y-prosemirror/src/plugins/sync-plugin.js:1125:20
at transact (/home/node/app/node_modules/yjs/dist/yjs.cjs:3415:14)
at Doc.transact (/home/node/app/node_modules/yjs/dist/yjs.cjs:562:12)
</code>
<br>

Also, as tiptap is based on prosemirror, if you search on google "prosemirror run on node js", you will find many things you can try, I am also trying them currently, will update here, if i found something.

@patrickalima98  were you able to resolve the issue & get the setup working?, If yes, can you share what worked for you, did you used something different?

**Reply by Jkense** [reply]
Any update on this?

**Reply by Taelar** [reply]
It seems to work with my nestJS project, just needed to add
`global.navigator = window.navigator`

**Comment by nperez0111** [comment, marked as answer]
It definitely is not supported to try to run tiptap on the server side. There are way too many DOM APIs that we rely on that would make it close to impossible to maintain. 

The recommended way to manipulate a document on the server side is to represent your content as JSON and from there you can use raw json manipulation. We are thinking of creating a package to help with common transforms you might need like the editor provides currently but have not gotten around to it yet
